src/maxtree.jl

```diff
@@ -0,0 +1,881 @@
+[`diameters`](@ref), [`area_opening`](@ref), [`area_closing`](@ref).
+"""
+function areas(maxtree::MaxTree)
+    areas = fill(1, length(maxtree)) # start with 1-pixel areas
+    parents = maxtree.parents
+    @inbounds for p in Iterators.Reverse(maxtree.traverse)
+        q = parents[p]
+        (q != p) && (areas[q] += areas[p])
+    end
+    return areas
+end
+
+"""
+    boundingboxes(maxtree::MaxTree) -> Matrix{Int}
+
+Computes the minimal bounding boxes of all `maxtree` components.
+
+# Returns
+The matrix, where the `i`-th column encodes the bounding box of the subtree
+with the root at the `i`-th pixel. The first ``N`` elements of the column
+define the minimal cartesian index of the bounding box, the last ``N``
+elements are its maximal cartesian index.
+
+# See also
+[`diameters`](@ref).
+"""
+function boundingboxes(maxtree::MaxTree{N}) where N
+    # initialize bboxes
+    bboxes = Matrix{Int}(undef, 2N, length(maxtree))
+    @inbounds for (i, ci) in enumerate(CartesianIndices(maxtree.axes))
+        offset = 2N * (i-1)
+        bboxes[offset .+ (1:N)] .= Tuple(ci)
+        bboxes[offset .+ ((N+1):2N)] .= Tuple(ci)
+    end
+
+    @inbounds for p in Iterators.Reverse(maxtree.traverse)
+        q = maxtree.parents[p]
+        for i in 1:N
+            bboxes[i, q] = min(bboxes[i, q], bboxes[i, p])
+        end
+        for i in (N+1):2N
+            bboxes[i, q] = max(bboxes[i, q], bboxes[i, p])
+        end
+    end
+    return bboxes
+end
+
+"""
+    diameters(maxtree::MaxTree) -> Vector{Int}
+
+Computes the "diameters" of all `maxtree` components.
+
+"Diameter" of the *max-tree* connected component is the length of
+the widest side of the component's bounding box.
+
+# Returns
+The `i`-th element of the result is the "diameter" of the `i`-th component.
+
+# See also
+[`boundingboxes`](@ref), [`areas`](@ref),
+[`diameter_opening`](@ref), [`diameter_closing`](@ref).
+"""
+diameters(maxtree::MaxTree{N}) where N =
+    dropdims(mapslices(boundingboxes(maxtree), dims=1) do bbox
+        maximum(ntuple(i -> bbox[i+N] - bbox[i] + 1, Val{N}()))
+    end, dims=1)
+
+"""
+    direct_filter!(output::AbstractArray, image::AbstractArray,
+                   maxtree::MaxTree, attrs::AbstractVector, min_attr) -> output
+
+Applies a direct filtering of the `image` and stores the result in `output`.
+
+Produces an image in which all components of its max-tree representation
+have the specified attribute value no less than `min_attr`. It's done by
+replacing the pixels of the component that has smaller attribute value with
+the value of the reference pixel of its first valid ancestor.
+
+# Arguments
+- `maxtree::MaxTree{N}`: pre-built max-tree of the `image`
+- `attrs::AbstractVector`: `attrs[i]` is the attribute value for the ``i``-th
+   component of the tree (``i`` being the linear index of its *reference pixel*)
+- `all_below_min`: the value to fill the `output` if all attributes of all
+  components (including the root one) are below `min_attr`
```

Comment:
It's still not exactly clear what this function does. Maybe a demo?

Reply:
I've renamed it `filter_components!()` (the original name is a direct copy from skimage) and changed the phrasing a bit.
In fact, this function is not exported (one good reason is that it requires that `attrs` is monotone wrt tree hierarchy, which would require checking when `attr` is supplied by a user).
So I think a demo is not necessary. I still decided to document it as a docstring, because it is the basis for all maxtree operations.